bundles/org.openhab.binding.revogi/src/main/java/org/openhab/binding/revogi/internal/udp/UdpSenderService.java

```diff
@@ -0,0 +1,143 @@
+        this.datagramSocketWrapper = datagramSocketWrapper;
+    }
+
+    public CompletableFuture<List<UdpResponseDTO>> broadcastUdpDatagram(String content) {
+        List<String> allBroadcastAddresses = NetUtil.getAllBroadcastAddresses();
+        CompletableFuture<List<UdpResponseDTO>> future = new CompletableFuture<>();
+        scheduler.submit(() -> future.complete(allBroadcastAddresses.stream().map(address -> {
+            try {
+                return sendMessage(content, InetAddress.getByName(address));
+            } catch (UnknownHostException e) {
+                logger.warn("Could not find host with IP {}", address);
+                return new ArrayList<UdpResponseDTO>();
+            }
+        }).flatMap(Collection::stream).distinct().collect(toList())));
+        return future;
+    }
+
+    public CompletableFuture<List<UdpResponseDTO>> sendMessage(String content, String ipAddress) {
+        try {
+            CompletableFuture<List<UdpResponseDTO>> future = new CompletableFuture<>();
+            InetAddress inetAddress = InetAddress.getByName(ipAddress);
+            scheduler.submit(() -> future.complete(sendMessage(content, inetAddress)));
+            return future;
+        } catch (UnknownHostException e) {
+            logger.warn("Could not find host with IP {}", ipAddress);
+            return CompletableFuture.completedFuture(Collections.emptyList());
+        }
+    }
+
+    private List<UdpResponseDTO> sendMessage(String content, InetAddress inetAddress) {
+        logger.debug("Using address {}", inetAddress);
+        byte[] buf = content.getBytes(Charset.defaultCharset());
+        DatagramPacket packet = new DatagramPacket(buf, buf.length, inetAddress, REVOGI_PORT);
+        List<UdpResponseDTO> responses = Collections.emptyList();
+        try {
+            datagramSocketWrapper.initSocket();
+            datagramSocketWrapper.sendPacket(packet);
+            responses = getUdpResponses();
+        } catch (IOException e) {
+            logger.warn("Error sending message or reading anwser {}", e.getMessage());
+        } finally {
+            datagramSocketWrapper.closeSocket();
+        }
+        return responses;
+    }
+
+    private List<UdpResponseDTO> getUdpResponses() {
+        int timeoutCounter = 0;
+        List<UdpResponseDTO> list = new ArrayList<>();
+        while (timeoutCounter < MAX_TIMEOUT_COUNT) {
+            byte[] receivedBuf = new byte[512];
+            DatagramPacket answer = new DatagramPacket(receivedBuf, receivedBuf.length);
+            try {
+                datagramSocketWrapper.receiveAnswer(answer);
+            } catch (SocketTimeoutException | SocketException e) {
+                timeoutCounter++;
+                try {
+                    TimeUnit.MILLISECONDS.sleep(timeoutCounter * timeoutBaseValue);
+                } catch (InterruptedException ex) {
+                    logger.debug("Interrupted sleep");
+                    Thread.currentThread().interrupt();
```

Comment:
An interrupt usually means that the thread should exit, so you should terminate as quick as possible instead of continuing your timeout loop.

Reply:
In that case it's the sleep thread that would be interrupted. I see no reason to not continue the loop itself.

Reply:
An interrupt is less a "get back to work" and more of a "stop all work and go home".

Reply:
I agree with @cpmeister. A good approach to finish the thread is to simply throw the exception and handle it top level. Another option would be to return an empty list.

The thread is interrupted e.g. when the JVM is shutdown.

Reply:
it's now implemented in a way it returns everything we have so far.